train/train.py

```diff
@@ -284,13 +308,50 @@ def train():
 
     random.shuffle(raw_data)
 
-    dataset = CustomDataset(raw_data, tokenizer)
+    if training_args.do_eval:
+        # Take 90:10 train-validation split
+        raw_train_data, raw_eval_data = split_data(raw_data, data_args.data_path, 0.9)
+        train_dataset = CustomDataset(raw_train_data, tokenizer)
+        eval_dataset = CustomDataset(raw_eval_data, tokenizer)
+    else:
+        train_dataset = CustomDataset(raw_data, tokenizer)
+        
+    def preprocess_logits_for_metrics(logits, labels):
+        pred_ids = torch.argmax(logits, dim=-1)
+        shift_logits = logits[..., :-1, :].contiguous()
+        shift_labels = labels[..., 1:].contiguous()
+        loss_fct = CrossEntropyLoss(reduction="none")
+        shift_logits = shift_logits.view(-1, tokenizer.vocab_size)
+        shift_labels = shift_labels.view(-1)
+        loss = loss_fct(shift_logits, shift_labels)
+        loss = torch.mean(loss.view(logits.shape[0], -1), dim=-1)
+        return pred_ids, loss
+    
+    def compute_metrics(eval_preds):
+        predictions = eval_preds.predictions[0][:, : -1]  # B x L
+        labels = eval_preds.label_ids[:, 1:]  #  B x L
+        acc_count = 0
+        total_num = 0
+        # Calculate accuracy
+        for pred, label in zip(predictions.flatten().tolist(), labels.flatten().tolist()):
+            if label != -100:
+                if label == pred:
+                    acc_count += 1
+                total_num += 1
+        # Calculate perplexity
+        loss = eval_preds.predictions[1].tolist()
+        loss = sum(loss) / len(loss)
+        perplexity = math.exp(loss)
+        return {"accuracy": acc_count / total_num, "perplexity": perplexity}
 
     trainer = Trainer(
         model=model,
         tokenizer=tokenizer,
         args=training_args,
-        train_dataset=dataset,
+        train_dataset=train_dataset,
+        eval_dataset=eval_dataset if training_args.do_eval else None,
+        compute_metrics=compute_metrics if training_args.do_eval else None,
+        preprocess_logits_for_metrics=preprocess_logits_for_metrics if training_args.do_eval else None,
```

Comment:
This whole one liner if else statements look complicated. Can you find a better way? If not, we can keep it.